docs/design/repair-with-hashes.md

```diff
@@ -0,0 +1,79 @@
+
+## Design
+
+When a satellite downloads pieces for repair, the storage node should also include the `PieceHash` and `OrderLimit` to the satellite along with each piece.
+
+The satellite should attempt to download exactly the Reed-Solomon minimum number of pieces during a repair, storing them in memory. It should validate the `OrderLimit` using its own private key.  It should validate that the `PieceHash` `Signature` corresponds to the `PublicKey` contained in the `OrderLimit`.  The satellite should calculate the size and hash of the piece as it downloads them, and it should be equal to the `PieceHash` `Hash` and `PieceSize`. If a piece fails any of these checks, an error should be logged and a different piece should be downloaded.  Once all pieces are downloaded, the `PiecePublicKey` of all pieces should be compared and equal.  If they are not equal, an error should be logged and more pieces should be downloaded until a minimum number of pieces share the same `PiecePublicKey`.  When enough pieces have been downloaded and validated, Reed-Solomon decoding should be used to reassemble the segment, and repair should continue as previously designed.
+
+Failed hash, signature, or public key checks may indicate cheating. Therefore audit must implement this same piece hash check logic. Changes to audit and the reporting of cheaters are otherwise outside the scope of this document.  
+
+## Rationale
+
+Both download and Reed Solomon decoding create more load for a satellite than hashing, so hashing should be preferred over downloading even a single piece.
+
+Downloading for repair is significantly different enough from streaming as to warrant a new, custom implementation.
+
+Using only the minimum number of pieces means that Reed-Solomon does not act as a check during repair. Hence hashing is used instead. While an uplink could potentially send signed bogus data to a storage node, the storage node would not be penalized by these actions. This requires that Audit implements a similar piece hash check instead of relying solely on Reed-Solomon encoding.
+
+The size of all piece hashes downloaded should be roughly equal to a default maximum segment size : 64MiB.  It seems preferable to keep this in memory over dealing with persistance to disk.
+
+## Implementation
+
+1. Add an optional `PieceHash` and `OrderLimit` fields to the `PieceDownloadResponse` protocol buffer for returning hashes.
+
+```message PieceDownloadResponse {
+    // Chunk response for download request
+    message Chunk {
+        int64 offset = 1;
+        bytes data = 2;
+    }
+    Chunk chunk = 1;
+    optional PieceHash hash = 2;
+    optional OrderLimit limit = 3;
+}```
+
+2. Alter the storage node code to populate `PieceHash` and `OrderLimit` when the `OrderLimit` `Action` is `GET_REPAIR`.
```

Comment:
It does not make sense to populate the `PieceHash` and `OrderLimit` with every chunk. Perhaps, populating them in the first-chunk message only would be enough.